### ISSUE TITLE
defer.fail breaks on empty value

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3145]
|<img alt="lightyear's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| lightyear reported|
|-|-|
|Trac ID|trac#3145|
|Type|defect|
|Created|2008-04-01 11:17:08Z|
|Branch|https://github.com/twisted/twisted/tree/defer-fail-3145|

If you run defer.fail without any parameters the API tells you that it works but it throws an exception:
```
Traceback (most recent call last):
  File "test_defer.py", line 283, in testImmediateFailure
    
  File "/home/ben/dev/twisted_trunk/twisted/internet/defer.py", line 64, in fail
    result = failure.Failure()
  File "/home/ben/dev/twisted_trunk/twisted/python/failure.py", line 173, in __init__
    raise NoCurrentExceptionError()
twisted.python.failure.NoCurrentExceptionError: 
```

I added a test that triggers that bug. Here is the patch
```
Index: twisted/test/test_defer.py
===================================================================
--- twisted/test/test_defer.py  (revision 23159)
+++ twisted/test/test_defer.py  (working copy)
@@ -280,6 +280,10 @@
         d.addErrback(l.append)
         self.assertEquals(str(l[0].value), "fail")

+        d = defer.fail()
+        d.addErrback(l.append)
+        self.assertEqulas(len(l), 2)
+
     def testPausedFailure(self):
         l = []
         d = defer.fail(GenericError("fail"))
```

<details><summary>Searchable metadata</summary>

```
trac-id__3145 3145
type__defect defect
reporter__lightyear lightyear
priority__low low
milestone__ 
branch__branches_defer_fail_3145 branches/defer-fail-3145
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1207048628000000 1207048628000000
changetime__1207063298000000 1207063298000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_1"></a>|@therve commented|
|-|-|

That's expected. The value is automatically retrieved if you are inside an except block. That's probably just a documentation issue.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [23162]) Branching to 'defer-fail-3145'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set owner to @exarkun|
|-|-|

Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set status to `closed`|
|-|-|

(In [23165]) Merge defer-fail-3145

Author: exarkun
Reviewer: therve
Fixes [#3145](https://github.com/twisted/twisted/issues/3145)

Clarify the documentation of twisted.internet.defer.fail and
twisted.internet.defer.Deferred.errback for the case where no
argument is supplied and there is no current exception state
(an exception will be raised).  Also simplify the implementation
of fail slightly, removing some duplication it had with errback.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
